src/biguint.rs

```diff
@@ -1749,6 +1749,50 @@ impl BigUint {
         }
         acc
     }
+
+    /// Finds square root of `self`.
+    ///
+    /// The result is the greatest integer less than or equal to the
+    /// square root of `self`.
+    pub fn sqrt(&self) -> Self {
+        let one = BigUint::one();
+        let two = BigUint::from(2 as u8);
+
+        // Trivial cases
+        if self.is_zero() {
+            return BigUint::zero();
+        }
+
+        if self.is_one() {
+            return one;
+        }
+
+        // Newton's method to compute the square root of an integer.
+        //
+        // Reference:
+        // Brent & Zimmermann, Modern Computer Arithmetic, v0.5.9, Algorithm 1.13
+        //
+        // Set initial guess to something >= floor(sqrt(self)), but as low
+        // as possible to speed up convergence.
+        let bit_len = self.len() * big_digit::BITS;
+        let guess = one.shl(bit_len/2 + 1);
+
+        let mut u = guess;
+        let mut s: BigUint;
+
+        loop {
+            s = u;
+            let q = self.div_floor(&s);
+            let t: BigUint = &s + &q;
```

Comment:
We don't need `q` after this, so adding `&s + q` by value will let it reuse that memory allocation.

Reply:
Indeed, fixed.